pkg/reconciler/v1alpha1/taskrun/resources/pre_post_build_step.go

```diff
@@ -0,0 +1,126 @@
+/*
+Copyright 2018 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resources
+
+import (
+	"path/filepath"
+
+	"fmt"
+
+	"github.com/knative/build-pipeline/pkg/apis/pipeline/v1alpha1"
+	buildv1alpha1 "github.com/knative/build/pkg/apis/build/v1alpha1"
+	"go.uber.org/zap"
+	corev1 "k8s.io/api/core/v1"
+)
+
+var (
+	pvcDir       = "/pvc"
+	workspaceDir = "/workspace"
+)
+
+// WrapPostBuildSteps adds steps to run after sources and before steps to override the source with passedconstraint source
+// also adds step to collect in the end to collect
+func WrapPostBuildSteps(
+	taskRun *v1alpha1.TaskRun,
+	b *buildv1alpha1.Build,
+	logger *zap.SugaredLogger,
+) {
+	var postBuildSteps = make(map[string][]string)
+	for _, postStep := range taskRun.Spec.PostBuiltSteps {
+		postBuildSteps[postStep.Name] = postStep.Paths
+	}
+
+	for _, source := range b.Spec.Sources {
+		if paths, ok := postBuildSteps[source.Name]; ok {
+			var newSteps []corev1.Container
+			for _, path := range paths {
+				var sPath string
+				if source.TargetPath == "" {
+					sPath = workspaceDir
+				} else {
+					sPath = filepath.Join(workspaceDir, source.TargetPath)
+				}
+				newSteps = append(newSteps, []corev1.Container{{
+					Name:         fmt.Sprintf("source-copy-make-%s", source.Name),
+					Image:        "ubuntu",
+					Command:      []string{"/bin/mkdir"},
+					Args:         []string{"-p", path},
+					VolumeMounts: []corev1.VolumeMount{getPvcMount(taskRun.Spec.PVCName)},
+				}, {
+					Name:         fmt.Sprintf("source-copy-%s", source.Name),
+					Image:        "ubuntu",
+					Command:      []string{"/bin/cp"},
```

Comment:
genuine naive question : `cp` or `rsync` ? :yum:

Reply:
`cp`. I did consider rsync but from my understanding it seems to be a better tool if parts of files are updated instead of complete set of files.